raiden/tests/integration/test_regression.py

```diff
@@ -36,9 +50,124 @@ def test_regression_unfiltered_routes(raiden_network, token_addresses, settle_ti
         app.raiden.poll_blockchain_events()
 
     transfer = app0.raiden.mediated_transfer_async(
-        token,
-        1,
-        app4.raiden.address,
-        1,
+        token_address=token,
+        amount=1,
+        target=app4.raiden.address,
+        identifier=1,
     )
     assert transfer.wait()
+
+
+@pytest.mark.parametrize('number_of_nodes', [3])
+@pytest.mark.parametrize('channels_per_node', [CHAIN])
+@pytest.mark.parametrize('blockchain_type', ['tester'])
+def test_regression_revealsecret_after_secret(raiden_network, token_addresses):
+    """ A RevealSecret message received after a Secret message must be cleanly
+    handled.
+    """
+    app0, app1, app2 = raiden_network
+    token = token_addresses[0]
+
+    identifier = 1
+    transfer = app0.raiden.mediated_transfer_async(
+        token_address=token,
+        amount=1,
+        target=app2.raiden.address,
+        identifier=identifier,
+    )
+    assert transfer.wait()
+
+    all_logs = app1.raiden.transaction_log.get_events_in_block_range(
+        0,
+        app1.raiden.get_block_number(),
+    )
+
+    secret = None
+    for log in all_logs:
+        event = log.event_object
+        if isinstance(event, SendRevealSecret):
+            secret = event.secret
+            break
+    assert secret
+
+    reveal_secret = RevealSecret(secret)
+    app2.raiden.sign(reveal_secret)
+
+    reveal_data = reveal_secret.encode()
+    app1.raiden.protocol.receive(reveal_data)
+
+
+@pytest.mark.parametrize('number_of_nodes', [2])
+@pytest.mark.parametrize('channels_per_node', [CHAIN])
+@pytest.mark.parametrize('blockchain_type', ['tester'])
+def test_regression_multiple_revealsecret(raiden_network, token_addresses):
+    """ Multiple RevealSecret messages arriving at the same time must be
```

Comment:
This is a very nice explanation of the race condition. Can you copy paste the entire text also to the PR's description?